src/testing.jl

```diff
@@ -109,6 +109,7 @@ function Base.unsafe_read(io::AsyncIO, to::Ptr{UInt8}, nb::UInt)
             splice!(buffer, 1:n)
             written += n
         end
+        eof(io) && throw(EOFError())
```

Comment:
```suggestion
        istaskdone(io.task) && throw(EOFError())
```

The second part of `eof` checks `@lock io.lock isempty(io.buffer)`, which may be false if we are waiting for new bytes to be written before copying the whole `nb` bytes at once.